server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java

```diff
@@ -42,11 +42,22 @@
     private ObjectObjectHashMap<String, CollectionStatistics> fieldStatistics = HppcMaps.newNoNullKeysMap();
     private int maxDoc;
 
-    public DfsSearchResult() {
-    }
-
     public DfsSearchResult(StreamInput in) throws IOException {
-        readFrom(in);
+        super(in);
+        requestId = in.readLong();
+        int termsSize = in.readVInt();
+        if (termsSize == 0) {
+            terms = EMPTY_TERMS;
+        } else {
+            terms = new Term[termsSize];
+            for (int i = 0; i < terms.length; i++) {
+                terms[i] = new Term(in.readString(), in.readBytesRef());
+            }
+        }
+        this.termStatistics = readTermStats(in, terms);
+        readFieldStats(in, fieldStatistics);
```

Comment:
nit: a bit unrelated, but looks like we are always passing in an empty HppcMaps.newNoNullKeysMap() here which otherwise gets create in this method. This could be changed so that readFieldStats takes only the input stream argument and always creates its own new HppcMaps.newNoNullKeysMap()  instance which it then returns. Might just be me but I find that clearer.